### PR TITLE
Added What's New orchestration hook

### PR DESCRIPTION
**Pull request by rob-ghost**
ref https://linear.app/ghost/issue/BER-2610

## Why are we making this change?

This PR completes the React implementation of the What's New feature by orchestrating the changelog data and user preferences to provide notification state. The core challenge is determining whether the current user has seen the latest changelog entry, and providing a simple API for UI components to display notifications and mark entries as seen.

This is the final piece in the What's New feature port from Ember to React.

## What does this PR do?

This PR introduces two hooks: `useWhatsNew` and `useDismissWhatsNew`.

The query hook combines the changelog data with the user's stored `lastSeenDate` preference to compute whether there are new entries. Rather than checking every changelog entry, we only need to compare the latest entry's publication date with when the user last viewed the changelog. This keeps the logic simple and efficient.

The hook also distinguishes between regular new entries and featured entries, since featured entries deserve special UI treatment (like more prominent badges or notifications). This distinction is computed by checking both the "new" status and the entry's featured flag.

For first-time users who have never opened What's New before, the hook automatically initializes their preferences with the current date. This prevents existing changelog entries from appearing as "new" when someone opens the feature for the first time.

The mutation hook provides a simple way to dismiss What's New notifications. When called, it updates the user's `lastSeenDate` to match the latest changelog entry's publication date, which automatically causes the query hook to return `hasNew: false` on subsequent renders.

This PR also refactors the user preferences system to use Zod codecs with explicit encode/decode functions. This allows the internal representation to use native `Date` objects for type safety, while automatically converting to/from ISO datetime strings for storage. This makes working with dates in the preferences system more ergonomic and less error-prone.

## Why is this needed?

The What's New feature needs a centralized source of truth for notification state. UI components shouldn't need to understand the relationship between changelog data and user preferences, or how to compute "new" status. By encapsulating this logic in hooks, we make it easy for multiple components to consistently show or hide notification badges, and we ensure the feature behaves correctly for both new and returning users.

The separate `hasNewFeatured` flag lets us create better UX by highlighting particularly important updates differently from routine changelog entries.

## Testing

```bash
yarn workspace @tryghost/admin test:unit src/whats-new/hooks/use-whats-new.test.tsx
```

Tests cover new user initialization, computing new entry status, the dismiss mutation, and handling edge cases like empty changelogs.

## Dependencies

Depends on:
- #25356 (test infrastructure + user preferences)
- #25357 (changelog hook)

### CODE REVIEW

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+    const parsedSettings = useMemo(() => parseWhatsNewSettings(preferences), [preferences]);
+
+    const settings = useMemo(() => parsedSettings ?? getDefaultWhatsNewSettings(), [parsedSettings]);
+
+    // Initialize default settings if missing
+    useEffect(() => {
+        if (!parsedSettings && isPreferencesLoaded) {
+            const defaultSettings = getDefaultWhatsNewSettings();
+            void updatePreferences({
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.ts:36]: Look into multiple side-effects in this case.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+const whatsNewQueryKey = (preferences: Preferences | undefined) => ["whatsNew", preferences?.whatsNew] as const;
+
+export const useWhatsNew = (): UseQueryResult<WhatsNewData> => {
+    const { data: preferences, isSuccess: isPreferencesLoaded } = useUserPreferences();
+    const { data: changelog, isSuccess: isChangelogLoaded } = useChangelog();
+    const { mutateAsync: updatePreferences } = useEditUserPreferences();
+
+    const parsedSettings = useMemo(() => parseWhatsNewSettings(preferences), [preferences]);
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.ts]: Factor out `parseWhatsNewSettings` into a `hasSettings` boolean now that we have Zod on preferences.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            const hasNew = latestEntry.publishedAt > settings.lastSeenDate;
+            const hasNewFeatured = hasNew && latestEntry.featured === true;
+
+            return { hasNew, hasNewFeatured };
+        },
+        enabled: isPreferencesLoaded && isChangelogLoaded,
+    });
+};
+
+export const useMarkWhatsNewAsSeen = (): UseMutationResult<void, Error, void, unknown> => {
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.ts]: Think of a better name for this, `useDismissWhatsNew` perhaps.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    return {
+        lastSeenDate: new Date(preferences.whatsNew.lastSeenDate),
+    };
+}
+
+function serializeWhatsNewSettings(whatsNewSettings: WhatsNewSettings) {
+    return {
+        lastSeenDate: whatsNewSettings.lastSeenDate.toISOString(),
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.ts]: Use [Zod codecs](https://zod.dev/codecs) for encode / decode of the User Preferences

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+function parseWhatsNewSettings(preferences: Preferences | undefined): WhatsNewSettings | null {
+    if (!preferences?.whatsNew?.lastSeenDate) {
+        return null;
+    }
+
+    return {
+        lastSeenDate: new Date(preferences.whatsNew.lastSeenDate),
+    };
+}
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.ts]: Move to preferences Zod schema

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+const whatsNewQueryKey = (preferences: Preferences | undefined) => ["whatsNew", preferences?.whatsNew] as const;
+
+export const useWhatsNew = (): UseQueryResult<WhatsNewData> => {
+    const { data: preferences, isSuccess: isPreferencesLoaded } = useUserPreferences();
+    const { data: changelog, isSuccess: isChangelogLoaded } = useChangelog();
+    const { mutateAsync: updatePreferences } = useEditUserPreferences();
+
+    const parsedSettings = useMemo(() => parseWhatsNewSettings(preferences), [preferences]);
+
+    const settings = useMemo(() => parsedSettings ?? getDefaultWhatsNewSettings(), [parsedSettings]);
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.ts]: Check if we need to `getDefaultWhatsNewSettings` if we're doing also in the side-effect.

File: apps/admin/src/whats-new/hooks/use-whats-new.test.tsx [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+        http.get(USERS_API_URL, () => {
+            return HttpResponse.json({
+                users: [
+                    {
+                        ...mockUser,
+                        accessibility: accessibility ?? mockUser.accessibility,
+                    },
+                ],
+            });
+        }),
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.test.tsx:103]: Can we compose this from the existing user preferences tests?

File: apps/admin/src/whats-new/hooks/use-whats-new.test.tsx [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+        // Mock the changelog endpoint
+        http.get(CHANGELOG_API_URL, () => {
+            return HttpResponse.json({
+                posts: changelog.posts ?? [],
+                changelogUrl: changelog.changelogUrl ?? "https://ghost.org/changelog",
+            });
+        })
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.test.tsx:123]: Can we compose this from the changelog tests?

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 1 above)
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts:36]: I spent a long time on this one only to come to the conclusion that the solutions were too complex to warrant doing. We're better off living with the potential race condition.

For reference, the best solution was to have a `WhatsNewProvider` parent component with the side effect of initialising before rendering children which use `useWhatsNew`.

File: apps/admin/src/whats-new/hooks/use-whats-new.test.tsx (hunk 7 above)
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.test.tsx:103]: After looking into it I decided that leaving the fixtures separate was best to avoid coupling the discrete test cases given their subtle differences.

File: apps/admin/src/whats-new/hooks/use-whats-new.test.tsx (hunk 8 above)
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.test.tsx:123]: After looking into it I decided that leaving the fixtures separate was best to avoid coupling the discrete test cases given their subtle differences.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            }
+
+            const latestEntry = entries[0];
+            const hasNew = latestEntry.publishedAt > whatsNewPreferences.lastSeenDate;
+            const hasNewFeatured = hasNew && latestEntry.featured === true;
+
+            return { hasNew, hasNewFeatured };
+        },
+        enabled: isPreferencesLoaded && isChangelogLoaded && hasWhatsNewPreferences,
+    });
```
[the remark this answers is not in the document]
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts]: This wasn't a concern because `changelog` is never invalidated, it has a `staleTime: infinity` and is intended to be fetched only once per app load.

That said, the dependant query should follow better conventions and include some aspect of the changelog in the query key to hint that it _should_ re-compute, even if it _doesn't_ in practical terms.

I've since updated `whatsNewQueryKey` to take an optional `ChangelogEntry` and we'll depend on its `publishedAt` date. If the latest entries `publishedAt` date changes, then we'll invalidate the key and re-compute the query result (though this won't happen in reality).

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            }
+
+            const hasNew = latestEntry.publishedAt > whatsNewPreferences.lastSeenDate;
+            const hasNewFeatured = hasNew && latestEntry.featured === true;
+
+            return { hasNew, hasNewFeatured };
+        },
+        enabled: isPreferencesLoaded && isChangelogLoaded && hasWhatsNewPreferences,
+        staleTime: Infinity,
+        cacheTime: 0,
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.ts:62]: This pattern rears its head again. I opted to put the options here for now, but we can consider encapsulating this in a `useLiveQuery` or `useDependantQuery` which curries these into `useQuery`. We'd have to decide what to call it and where to keep it, but the idea of a query derived from another query and the whole point is to not keep cached values and never go stale makes sense as a re-useable hook.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    };
+}
+
+interface WhatsNewData {
+    hasNew: boolean;
+    hasNewFeatured: boolean;
+}
+
+const whatsNewQueryKey = (preferences: Preferences | undefined, latestEntry: ChangelogEntry | undefined) =>
+    ["whatsNew", preferences?.whatsNew?.lastSeenDate.toISOString(), latestEntry?.publishedAt.toISOString()] as const;
```
[the remark this answers is not in the document]
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts]: This is factually incorrect. The optional chain operator `?` will **early out** with `undefined` if any part of the chain is `undefined` / `null`, so we will never get to `. toISOString()`.

This is a pretty fundamental error in your javascript knowledge.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+export const useWhatsNew = (): UseQueryResult<WhatsNewData> => {
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.ts:26]: **Meta question:** Does "whats new" make sense as a name anymore?

`whats-new` refers to the title of the notification banner which is shown to the user and the service responsible for showing this banner was named `whats-new`.

In this new hooks landscape we've decomposed the service into many hooks, and now this `useWhatsNew` hook has inherited the name. However, if we look at the return values, `{ hasNew, hasNewFeatured }` the name `whats-new` no longer makes much sense.

Perhaps we should rename this to:

1. `useUnseenChangelog -> { hasUnseen, hasUnseenFeatured }`
2. `useMarkChangelogSeen`

I think these might be more accurate names, referring to unseen changelog entries and then "seeing" them.

What do we think?

File: apps/admin/src/hooks/user-preferences.ts [hunk 13]
```diff
@@ -5,15 +5,32 @@ import { useQueryClient } from "@tryghost/admin-x-framework";
...
+        decode: (stored) => ({
+            ...stored,
+            whatsNew: stored.whatsNew ? WhatsNewPreferencesCodec.decode(stored.whatsNew) : undefined,
+        }),
+        encode: (working) => ({
+            ...working,
+            whatsNew: working.whatsNew ? WhatsNewPreferencesCodec.encode(working.whatsNew) : undefined,
+        }),
+    }
+);
```
Comment (jonatansberg) [inline apps/admin/src/hooks/user-preferences.ts]: I think it would make more sense to put these codecs on the field themselves rather than the full schema. That way you don't have to reach in to objects or deal with the other existing fields when encoding or decoding.

```ts
const isoDatetimeToDate = z.codec(z.iso.datetime(), z.date(), {
    decode: (isoString) => new Date(isoString),
    encode: (date) => date.toISOString(),
});

const WhatsNewSchema = z.looseObject({ 
    lastSeenDate: isoDatetimeToDate
})

const PreferencesSchema = z.looseObject({ 
    whatsNew: WhatsNewSchema.optional()
});
```

File: apps/admin/src/hooks/user-preferences.ts [hunk 14]
```diff
@@ -30,7 +47,7 @@ export const useUserPreferences = (): UseQueryResult<Preferences> => {
             const raw = user.accessibility || "{}";
             const parsed = JSON.parse(raw) as unknown;
 
-            return PreferencesSchema.parse(parsed);
+            return PreferencesCodec.parse(parsed);
```
Comment (jonatansberg) [inline apps/admin/src/hooks/user-preferences.ts]: I don't think we should rename the schema. The codec thing is an implementation level concern of making the transforms bidirectional. It's still a schema.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (jonatansberg) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts]: Couldn't `lastSeenDate` potentially still being null (depending on how you define the schema) which would throw?

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 12 above)
Comment (jonatansberg) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts:26]: The "What's New" name makes sense for us internally and it's aligned with what is stored in the user preferences. I think it's good enough.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 1 above)
Comment (jonatansberg) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts:36]: You could also pull out the side effect into an initialize-what's-new-thing and only called that hook once 🤔 

Otherwise, I had something like this in mind:
https://tkdodo.eu/blog/concurrent-optimistic-updates-in-react-query

Did you look into anything like that?

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 10 above)
Comment (jonatansberg) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts:62]: Let's tackle this as a follow-up.

**Review by jonatansberg — CHANGES_REQUESTED** [review]
I think we might want to make some changes to the schema. Otherwise, this looks good.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts]: The schema claims it not to be null so either `whatsNew` is defined with a `lastSeenDate` or its not, there's no in-between.

IIRC, that's true of the existing data for users also based on how the current `whats-new` service is implemented.

There is a risk that there's some Users out there with invalid preferences and/or whats new preferences, and do we want to be accepting of those (and possibly sanitise them) or do we want to raise exceptions. How permissive do we want to be with this data?

Since we moved `whatsNew` validation into `user-preferences` its a little harder to have error handling / fallback behaviour for individual preference keys, either all the preferences are valid, or none of them are. I'll look into Zod supports defaulting a whole sub-schema in the case where the sub-schema fails to validate so we can handle sanitising `whatsNew`.

File: apps/admin/src/hooks/user-preferences.ts (hunk 13 above)
Comment (rob-ghost) [reply apps/admin/src/hooks/user-preferences.ts]: Much cleaner 👍🏻

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 1 above)
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts:36]: > You could also pull out the side effect into an initialize-what's-new-thing and only called that hook once 🤔

Can you give an example of what this looks like?

> Did you look into anything like that?

Yep one of the attempts used `cancelQueries` and `isMutating` but ultimately it was pretty complex to setup and test compared to accepting the race condition so much like the Provider solution, I decided against it. I could put together an example diff if you'd like to see the options before we accept the solution?

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts]: We will get loose with our user preferences schema, and add some logging to validation errors.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 1 above)
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts:36]: TODO after merge; check both the module-level variable lock and also lifting the initialisation into a separate hook which is run from `<App>` (or some other kind of top level context/component).

File: apps/admin/src/whats-new/hooks/use-whats-new.ts [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+            // Safe to assert non-null because query is only enabled when hasWhatsNewPreferences is true
+            const whatsNewPreferences = preferences!.whatsNew!;
+
+            const hasNew = latestEntry.publishedAt > whatsNewPreferences.lastSeenDate;
+            const hasNewFeatured = hasNew && latestEntry.featured === true;
+
+            return { hasNew, hasNewFeatured };
+        },
+        enabled: isPreferencesLoaded && isChangelogLoaded && hasWhatsNewPreferences,
```
Comment (rob-ghost) [inline apps/admin/src/whats-new/hooks/use-whats-new.ts]: `hasWhatsNewPreferences` makes `isPreferencesLoaded` redundant.

File: apps/admin/src/whats-new/hooks/use-whats-new.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (rob-ghost) [reply apps/admin/src/whats-new/hooks/use-whats-new.ts]: I made the schema members optional and also permissive of validation errors through `catch(undefined)`.

I tried to add logging but I didn't like mixing in `console.error` into each `catch` call, and even with a custom `catchAndLog` helper it would still be mixing concerns I think.

I tried `safeParse` which works but it meant we'd have to fallback to an empty object for all preferences which felt too nuclear just to get a log whenever we have potentially invalid preferences.

I felt catching individually met most of the requirements, it is permissive with defaults defined within the schema and doesn't mix parsing behaviour concerns.

I updated the rest of the code to handle the optional cases for all whats new preference fields.

**Review by jonatansberg — APPROVED** [review]
Looks good! There's a merge conflict that need resolving.
